tests/integration/integration_suite_test.go

```diff
@@ -1,13 +1,187 @@
-package integration_test
+package integration
 
 import (
+	. "github.com/k8ssandra/k8ssandra/tests/integration/steps"
+
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
 	"testing"
+)
 
-	. "github.com/onsi/ginkgo"
-	. "github.com/onsi/gomega"
+const (
+	medusaTestTable    = "medusa_test"
+	medusaTestKeyspace = "medusa"
 )
 
-func TestIntegration(t *testing.T) {
-	RegisterFailHandler(Fail)
-	RunSpecs(t, "Integration Suite")
+func TestMain(m *testing.M) {
+	updateHelmDeps()
+	os.Exit(m.Run())
+}
+
+func initializeCluster(t *testing.T) {
+	CheckK8sClusterIsReachable(t)
+	InstallTraefik(t)
+	CreateNamespace(t)
+	CheckK8ssandraNamespaceWasCreated(t)
+}
+
+func cleanupCluster(t *testing.T) {
+	UninstallK8ssandraHelmRelease(t)
+	UninstallTraefikHelmRelease(t)
+	DeleteK8ssandraNamespace(t)
+	DeleteTraefikNamespace(t)
+	CheckNamespaceIsAbsent(t)
+}
+
+// Reaper scenario:
+// - Install Traefik
+// - Create a namespace
+// - Register a cluster with 3 Cassandra nodes
+// - Verify that Reaper correctly initializes
+// - Start a repair on the reaper_db keyspace
+// - Wait for at least one segment to be processed
+// - Cancel the repair
+// - Terminate the namespace and delete the cluster
+func TestReaperDeploymentScenario(t *testing.T) {
+	t.Run("Initialize cluster", initializeCluster)
```

Comment:
I am not sure if makes sense to organize all of these as subtests. For instance, if `initializeCluster` fails, every test that follows will fail, so there is no point in running them in that case.

Reply:
GH wouldn't let me update my previous comment. It seems like it would make more sense to treat `initializeCluster` and `deployClusterForReaper` as setup. If either of those fail, there really is not much point in proceeding.

Reply:
They won't as we invoke the tests with the `-failfast` flag.
Organizing the steps in subtests gives us: 

- Readability for the test code
- Clean output with each subtest description being logged when it starts
- The ability to run single subtests when needed (for dev/troubleshooting purposes for example).

I'm -1 on not using subtests here since we can achieve the desired behavior and exit on failure.

Here's an example of a test failing in the "Initialize cluster" subtest:

```
% go test . -v -failfast -run=TestMedusaDeploymentScenario/"Minio:"
2021/03/17 08:38:28 Updating Helm dependencies...
=== RUN   TestMedusaDeploymentScenario
=== RUN   TestMedusaDeploymentScenario/Minio:_Initialize_cluster
TestMedusaDeploymentScenario/Minio:_Initialize_cluster 2021-03-17T08:38:34+01:00 logger.go:66: Running command kubectl with args [--namespace default get namespace default -o jsonpath={.metadata.name}]
TestMedusaDeploymentScenario/Minio:_Initialize_cluster 2021-03-17T08:38:34+01:00 logger.go:66: The connection to the server localhost:8080 was refused - did you specify the right host or port?
    TestMedusaDeploymentScenario/Minio:_Initialize_cluster: integration_steps.go:219:
        Expected
            <*shell.ErrWithCmdOutput | 0xc00027e040>: {
                Underlying: {
                    ProcessState: {
                        pid: 99393,
                        status: 256,
                        rusage: {
                            Utime: {
                                Sec: 0,
                                Usec: 40937,
                                Pad_cgo_0: [0, 0, 0, 0],
                            },
                            Stime: {
                                Sec: 0,
                                Usec: 20864,
                                Pad_cgo_0: [0, 0, 0, 0],
                            },
                            Maxrss: 24985600,
                            Ixrss: 0,
                            Idrss: 0,
                            Isrss: 0,
                            Minflt: 3428,
                            Majflt: 2903,
                            Nswap: 0,
                            Inblock: 0,
                            Oublock: 0,
                            Msgsnd: 0,
                            Msgrcv: 0,
                            Nsignals: 61,
                            Nvcsw: 29,
                            Nivcsw: 922,
                        },
                    },
                    Stderr: nil,
                },
                Output: {
                    stdout: {
                        Lines: nil,
                        merged: {
                            Mutex: {state: 0, sema: 0},
                            Lines: [
                                "The connection to the server localhost:8080 was refused - did you specify the right host or port?",
                            ],
                        },
                    },
                    stderr: {
                        Lines: [
                            "The connection to the server localhost:8080 was refused - did you specify the right host or port?",
                        ],
                        merged: {
                            Mutex: {state: 0, sema: 0},
                            Lines: [
                                "The connection to the server localhost:8080 was refused - did you specify the right host or port?",
                            ],
                        },
                    },
                    merged: {
                        Mutex: {state: 0, sema: 0},
                        Lines: [
                            "The connection to the server localhost:8080 was refused - did you specify the right host or port?",
                        ],
                    },
                },
            }
        to be nil
--- FAIL: TestMedusaDeploymentScenario (0.14s)
    --- FAIL: TestMedusaDeploymentScenario/Minio:_Initialize_cluster (0.14s)
FAIL
FAIL	github.com/k8ssandra/k8ssandra/tests/integration	6.177s
FAIL
```

You can see it stops after the first failure.

Reply:
>I'm -1 on not using subtests here since we can achieve the desired behavior and exit on failure.

It is not exactly the desired behavior. I believe `-failfast` halts execution of all non-parallel tests after the first failure. Consider this example:

```go
// example1_test.go

func TestFoo(t *testing.T) {
	t.Run("TEST ONE", testOne)
	t.Run("TEST TWO", testTwo)
	t.Run("TEST THREE", testThree)
}

func TestBar(t *testing.T) {

}

func testOne(t *testing.T) {
	t.Fatal("test one failed")
}

func testTwo(t *testing.T) {
	t.Fatal("test two failed")
}

func testThree(t *testing.T) {
}

// example2_test.go
func TestAnotherFoo(t *testing.T) {
}

func TestAnotherBar(t *testing.T) {
	
}
```

Here is the output when I run with `-failfast`:

```
go test -v -failfast -timeout 1800s ./test/e2e/...
=== RUN   TestFoo
=== RUN   TestFoo/TEST_ONE
    example1_test.go:16: test one failed
--- FAIL: TestFoo (0.00s)
    --- FAIL: TestFoo/TEST_ONE (0.00s)
FAIL
FAIL	github.com/k8ssandra/medusa-operator/test/e2e	0.367s
```
Not only are the remaining subtests in `TestFoo` skipped but `TestBar`, `TestAnotherBar` and `TestAnotherFoo` are skipped as well. I don't think we want this behavior.

Reply:
@jsanda In the case of integration tests, is it safe then to say that each test Go file will scope a single integration test run and will be run independent of the others?  What kind of resources will be required to do this effectively.  I'm thinking of things like snapshots of current k8ssandra state to determine root cause.  If I understand correctly, we would need to use appropriate test-name, namespace, context, etc.  

@adejanovski How would that work with your current approach of harvesting details after a test run?